[GEOS-10426] GSIP-167: Log4j 2 upgrade

jodygarnett opened the pull request:
[![GEOS-10426](https://badgen.net/badge/JIRA/GEOS-10426/0052CC)](https://osgeo-org.atlassian.net/browse/GEOS-10426)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

See proposal https://github.com/geoserver/geoserver/wiki/GSIP-167 - and prior https://github.com/geoserver/geoserver/pull/5734 pull request.

Other than messing with dependencies and user interface, the key technical component of this PR is creating GeoServerXmlConfiguration in order process the "Node" data structure produced by parsing the XML document before it is used to setup Log4J appenders and loggers. For a detailed discussion on limitations found and eventual approach used see https://issues.apache.org/jira/browse/LOG4J2-3469

Try this out as 2.21-M0 milestone release!

- https://github.com/geoserver/geoserver/releases/tag/2.21-M0
- https://sourceforge.net/projects/geoserver/files/GeoServer/2.21-M0/

Looking at the following:

* [x] [GEOS-10426](https://osgeo-org.atlassian.net/browse/GEOS-10426) GISP 167: Upgrade Log4j
* [x] [GEOS-9734](https://osgeo-org.atlassian.net/browse/GEOS-9734) Some classes use org.apache.log4j.Logger instead of org.geotools.util.logging.Logging.getLogger()
* [x] [GEOS-10241](https://osgeo-org.atlassian.net/browse/GEOS-10241) GetFeatureInfoJSONTest is unnecessarily verbose with system integration startup and logging

Related:

* https://github.com/geotools/geotools/pull/3809
* https://github.com/GeoWebCache/geowebcache/pull/1025
* https://github.com/geotools/geotools/pull/3870

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue https://osgeo-org.atlassian.net/browse/GEOS-10426
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

<!--Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.-->

jodygarnett commented:
There is one concerning windows failure, with a test case unable to clean up a temporary log file. Going to have to see if we can shut down harder or something so the files are released.

jodygarnett commented:
Can someone run formatter on this branch please; I am away from keyboard until May 1st.

aaime commented on src/main/src/test/java/org/geoserver/test/GeoServerAbstractTestSupport.java:
```diff
@@ -249,7 +249,7 @@ protected String[] getSpringContextLocations() {
      */
     protected String getLogConfiguration() {
-        return "/TEST_LOGGING.properties";
+        return "TEST_LOGGING.properties";
```
Oh so the tests are still using the property files? But in GeoServerSystemTestSupport names without extensions are used instead (so I'm guessing XML files are actually going to be used).

jodygarnett replied:
We could confirm with debugger; or just change this to ``TEST_LOGGING``.

aaime commented on src/main/src/test/java/org/geoserver/test/GeoServerBaseTestSupport.java:
```diff
@@ -136,7 +139,15 @@ public final void doSetup() throws Exception {
             testData = createTestData();
-            testData.setUp();
+
+            Logger imageLogger = Logger.getLogger("org.geotools.image");
```
Interesting, why the new addition? Isn't this out of line with the ability to choose a logging profile while writing tests?
I often override getLogConfiguration, when a test is failing, to get more information. But this one bit is static, would not follow the chosen logging profile.

jodygarnett replied:
If I remember it is due to an oddity of the java until logging which can be used when running some unit tests; if a logger is not instantiated, then the level is not available for child loggers.

aaime replied:
Would it allow customization of the logging level of "org.geotools.image" by overriding ``GeoSeverSystemTestSupport.getLogConfiguration``, or not?

jodygarnett replied:
sorry looking at the complete code example, this was to suppress endless info messages during setup before restoring the prior level.

jodygarnett replied:
So If I understand your feedback you would prefer changing TEST_LOGGING.xml to reduce the level for `org.geotools.image`; trouble is I only want to reduce the level during setup....

aaime commented on src/restconfig/src/test/java/org/geoserver/rest/LoggingControllerTest.java:
```diff
@@ -44,12 +45,14 @@ public void reset() throws Exception {
     public void testGetLoggingAsJSON() throws Exception {
         JSON json = getAsJSON(RestBaseController.ROOT_PATH + "/logging.json");
-        print(json);
+        if (LOGGER.isLoggable(Level.FINE)) {
```
Probably best leaving it as it was before...
Before: the print method internally checks for quiet tests, won't report anything during a build, but would dump info when running from IDE (when a developer most needs it).
After: dumps nothing regardless,  unless one also sets the logging profile to GEOSERVER_DEVELOPER_LOGGING. Seems overkill..

jodygarnett replied:
I would like to improve this method name to something more specific then, a method called "print" is just begging to be commented out or removed :)

aaime replied:
That is probably best discussed on the devel list, so that everyone at least knows about it. Those print methods have been there for something like 15 years (thinking about it sibling printing a XML DOM) with no other complaint.

aaime commented on src/web/app/src/main/webapp/WEB-INF/web.xml:
```diff
@@ -61,7 +61,15 @@
     </context-param>
-    
+
+    <!-- do not auto init log4j, geoserver will do it -->
```
Commented out bit... seems like you wanted to disable log4j auto init, but in the end it was no needed?

jodygarnett replied:
correct; was having a lot of trouble determining order of execution. Finally just put a breakpoint in ...

aaime reviewed (changes requested):
Here is a first pass of a review, based on code inspection. Will follow up with comments based on interactive testing later.

aaime commented:
Looking at the Windows failures:

```
2022-04-25T03:31:53.6993782Z 03:31:53,697 [ERROR] Tests run: 6, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 20.432 s <<< FAILURE! - in org.geoserver.rest.LoggingControllerTest
2022-04-25T03:31:53.6994953Z 03:31:53,697 [ERROR] org.geoserver.rest.LoggingControllerTest  Time elapsed: 19.411 s  <<< ERROR!
2022-04-25T03:31:53.6995596Z java.io.IOException: Failed to clean up test data dir after 100 attempts
2022-04-25T03:31:53.6996106Z Caused by: java.io.IOException: 
2022-04-25T03:31:53.6996493Z Failed to delete tree:
2022-04-25T03:31:53.6996870Z .\target\default2710106767194622034data
2022-04-25T03:31:53.6997204Z ???logs
2022-04-25T03:31:53.6997709Z    ???geoserver-test-2.log
2022-04-25T03:31:53.6998068Z    ???geoserver.log
2022-04-25T03:31:53.6998434Z    ???GEOSERVER_DEVELOPER_LOGGING.xml
2022-04-25T03:31:53.6998808Z    ???GEOTOOLS_DEVELOPER_LOGGING.xml
2022-04-25T03:31:53.6999185Z    ???PRODUCTION_LOGGING.xml
2022-04-25T03:31:53.6999534Z    ???QUIET_LOGGING.xml
2022-04-25T03:31:53.6999852Z    ???TEST_LOGGING.xml
2022-04-25T03:31:53.7000201Z    ???VERBOSE_LOGGING.xml
2022-04-25T03:31:53.7000453Z 
2022-04-25T03:31:53.7000905Z Caused by: org.apache.commons.io.IOExceptionList: 1 exceptions: [java.io.IOException: Unable to delete file: .\target\default2710106767194622034data\logs]
2022-04-25T03:31:53.7001596Z Caused by: java.io.IOException: Unable to delete file: .\target\default2710106767194622034data\logs
2022-04-25T03:31:53.7002127Z Caused by: java.nio.file.FileSystemException: 
2022-04-25T03:31:53.7002899Z .\target\default2710106767194622034data\logs\geoserver-test-2.log: The process cannot access the file because it is being used by another process.
```

```
2022-04-25T03:27:08.9521760Z 03:27:08,944 [ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 55.637 s <<< FAILURE! - in org.geoserver.web.admin.LogPageTest
2022-04-25T03:27:08.9523046Z 03:27:08,944 [ERROR] testNullLocation(org.geoserver.web.admin.LogPageTest)  Time elapsed: 30.58 s  <<< ERROR!
2022-04-25T03:27:08.9523844Z java.io.IOException: Failed to clean up test data dir after 100 attempts
2022-04-25T03:27:08.9527103Z Caused by: java.io.IOException: 
2022-04-25T03:27:08.9527736Z Failed to delete tree:
2022-04-25T03:27:08.9528123Z .\target\default4486438951823783280data
2022-04-25T03:27:08.9528467Z ???logs
2022-04-25T03:27:08.9528787Z    ???geoserver.log
2022-04-25T03:27:08.9554559Z    ???GEOSERVER_DEVELOPER_LOGGING.xml
2022-04-25T03:27:08.9555259Z    ???GEOTOOLS_DEVELOPER_LOGGING.xml
2022-04-25T03:27:08.9575487Z    ???PRODUCTION_LOGGING.xml
2022-04-25T03:27:08.9576055Z    ???QUIET_LOGGING.xml
2022-04-25T03:27:08.9576411Z    ???TEST_LOGGING.xml
2022-04-25T03:27:08.9576768Z    ???VERBOSE_LOGGING.xml
2022-04-25T03:27:08.9577038Z 
2022-04-25T03:27:08.9577505Z Caused by: org.apache.commons.io.IOExceptionList: 1 exceptions: [java.io.IOException: Unable to delete file: .\target\default4486438951823783280data\logs]
2022-04-25T03:27:08.9578215Z Caused by: java.io.IOException: Unable to delete file: .\target\default4486438951823783280data\logs
2022-04-25T03:27:08.9578756Z Caused by: java.nio.file.FileSystemException: 
2022-04-25T03:27:08.9579374Z .\target\default4486438951823783280data\logs\geoserver.log: The process cannot access the file because it is being used by another process.
```

The only two failures available in the logs are LogPageTest and LoggingControllerTest, so both seem related to manipulation of the log configuration... and the files that fail to be deleted are the ".log" files.
Thinking out loud: is it possible that while reconfiguring Log4J, the file appender from the previous config is left open?

jodygarnett commented on src/community/geomesa/pom.xml:
```diff
@@ -118,9 +118,17 @@ application directory.
-            <groupId>org.slf4j</groupId>
-            <artifactId>slf4j-api</artifactId>
-            <version>1.7.5</version>
+          <groupId>org.apache.logging.log4j</groupId>
+          <artifactId>log4j-api</artifactId>
+        </dependency>
+        <dependency>
+          <groupId>org.apache.logging.log4j</groupId>
+          <artifactId>log4j-core</artifactId>
+        </dependency>
```
The api to accept logging, and then core which actually does output.  

May be able to simplify if there is a transitive dependency.

jodygarnett commented on src/platform/src/main/java/org/geoserver/platform/GeoServerResourceLoader.java:
```diff
@@ -481,7 +481,7 @@ public void copyFromClassPath(String classpathResource, File target, Class<?> sc
                             + System.getProperty("user.name"));
         } catch (IOException e) {
-            LOGGER.log(Level.WARNING, "Error trying to copy logging configuration file", e);
+            // LOGGER.log(Level.CONFIG, "Unable to copy logging configuration file", e);
```
I think I actually found that the message was being logged twice; will need to double check.

aaime replied:
Seen the change thanks... why logging FINE level though? If a logging configuration file cannot be copied, then the rest of the logging configuration will not work (it expects to find the files, no?). Seems worth of a more severe indication (WARNING).

jodygarnett replied:
> Right after startup, Config UI does not show a selection of logging level, despite logging indicating the DEFAULT_LOGGING profile has been chosen (see above)

I was able to resolve this, on initial start value was set to "DEFAULT_LOGGING.properties" (and only "DEFAULT_LOGGING" was listed as an option in the UI)

jodygarnett commented on doc/en/developer/source/qa-guide/index.rst:
```diff
@@ -135,11 +180,33 @@ or if it's a general one that should be ignored, the `build/qa/spotbugs-exclude.
+   :end-before: </plugin>
+   :dedent: 12
+
 Any failure to comply with the rules will show up as a compiler error in the build output, e.g.::
```
not sure I follow? is the code example showing too much, or is the contents of the page to random? I was mostly taking notes when trying to figure out how to run things in isolation (while waiting for long builds).

aaime replied:
See here for a formatted version:
https://github.com/geoserver/geoserver/blob/log4j2_upgrade/doc/en/developer/source/qa-guide/index.rst

The title says "checkstyle" , yet you are suddenly changing topic to sortpom and spotless, which have nothing to do with checkstyle.

jodygarnett replied:
Got it, adding additional headings.

jodygarnett commented on src/community/jms-cluster/jms-geoserver/src/test/java/org/geoserver/cluster/integration/GeoServerInstance.java:
```diff
@@ -185,11 +185,7 @@ private GeoServerTestApplicationContext initInstance() throws Exception {
-                false,
-                true,
-                null);
+                loader, loader.get("/logs/TEST_LOGGING.xml"), false, true, null);
```
I could not get a classpath fallback to work; I even tried making a resource placeholder that would stream from the classpath (but ended up throwing it out has too complicated to live).

I changed to resource (from inputsteam) so I could check the name and stream content as needed.

jodygarnett replied:
The method ensures the built-in configurations are unpacked and present; so this code will run.

jodygarnett replied:
I tried out an example with the OL_LOGGING.properties just to confirm that this can be done safely.

jodygarnett commented on src/extension/wps/wps-core/src/main/java/org/geoserver/wps/UnsupportedParameterTypeProcessFilter.java:
```diff
@@ -52,7 +52,7 @@ public void setApplicationContext(ApplicationContext context) throws BeansExcept
                         if (ppios.isEmpty()) {
                             LOGGER.log(
-                                    Level.INFO,
+                                    Level.CONFIG,
```
This was an outstanding jira (to introduce CONFIG level logging). We can change default logging to include CONFIG (indeed this has been requested by a reviewer on the user list).

aaime replied:
The outstanding Jira was not part of the GSIP and I can see it added quite a bit of changes to this PR. To some extent, I understand doing more for the benefit of the community, but it should come with no regressions....

jodygarnett replied:
I will adjust DEFAULT_LOGGING to use CONFIG (so no loss of information being logged).

Related to "no regressions" target, I prefer to focus on log4j 2 best practice. Presently I have maintained the 3 plain text backups, where the log4j examples tend to compress backups to save space.

jodygarnett replied:
I have changed DEFAULT_LOGGING (and friends) to config level (rather than info level); so the same level of detail is maintained.

jodygarnett commented on src/extension/wps/wps-core/src/test/java/org/geoserver/wps/gs/UniqueProcessTest.java:
```diff
@@ -81,39 +80,29 @@ public void testUnique() throws Exception {
+        Logger logger = Logging.getLogger(JDBCDataStore.class);
+        Level previousLevel = logger.getLevel();
+        logger.info("hello world");
+        try (TestAppender appender = TestAppender.createAppender("testAutoCodeLogsErrors", null)) {
```
Yeah it was being too tough to debug half hearted attempts to check logs, so I made one good one.

jodygarnett commented on src/main/src/main/java/org/geoserver/filters/LoggingFilter.java:
```diff
@@ -32,6 +32,43 @@ public class LoggingFilter implements Filter {
+        String mimeType = sub == -1 ? contentType : contentType.substring(0, sub).toLowerCase();
+        String subType = sub == -1 ? "" : contentType.substring(sub + 1).toLowerCase();
+
+        if (mimeType.equals("text")) {
+            return false;
+        } else if (mimeType.equals("image")) {
+            if (subType.contains("svg")) {
+                return false;
+            }
+            return true;
+        } else if (mimeType.equals("application")) {
+            if (subType.equals("zip")) {
+                return true;
+            }
+            if (subType.contains("xml") || subType.contains("json") || subType.contains("gml")) {
+                return false;
+            }
+            return true;
+        } else {
+            return false; // assume text by default
+        }
```
yeah okay, I like that.

jodygarnett commented on src/main/src/main/java/org/geoserver/filters/LoggingFilter.java:
```diff
@@ -88,12 +125,18 @@ public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
                     req = new BufferedRequestWrapper(hreq, encoding, bytes);
+
+                    if (isBinary(hreq.getHeader("Content-type"))) {
+                        message += bytes.length + " bytes\n";
```
I think it is obvious from the logs, but we can indicated: message += bytes.length + " bytes (binary content)"

aaime replied:
I'm missing the obviousness... put yourself in the place of a user that hasn't seen the code writing the logs?

jodygarnett commented on src/main/src/main/java/org/geoserver/logging/LoggingUtils.java:
```diff
@@ -6,223 +6,726 @@
+            URI configLocation = Resources.file(configResource).toURI();
+
+            // Use log4j 1.2 API to reset
+            // org.apache.log4j.LogManager.resetConfiguration();
```
I *think* it can be removed; there is one user testing log4j 1.2 configuration on the user list I would like to hear back from.

jodygarnett commented on src/main/src/main/java/org/geoserver/GeoserverInitStartupListener.java:
```diff
@@ -434,6 +442,74 @@ public void contextDestroyed(ServletContextEvent sce) {
+                    Logging.ALL.setLoggerFactory("org.geotools.util.logging.LogbackLoggerFactory");
+                    break;
+                case Log4J2:
+                    Logging.ALL.setLoggerFactory("org.geotools.util.logging.Log4J2LoggerFactory");
```
Correct, the setLoggerFactory(String) is far more kind in the event of classpath requirements (indeed it is the reason for all the integration testing in geotools).

jodygarnett commented on src/main/src/main/resources/GEOSERVER_DEVELOPER_LOGGING.xml:
```diff
@@ -0,0 +1,68 @@
+<!-- during data_dir upgrades and in case the chosen logging config isn't available.            -->
+<Configuration name="DEFAULT_LOGGING" status="fatal" dest="out">
+    <CustomLevels>
+        <CustomLevel name="CONFIG" intLevel="450" />
```
That is unclear to me also; I did create the levels in geotools but as I understand it they static methods in geotools may not be called before log4j configuration file is loaded.

It may be okay now that I am creating GeoServerXMLConfiguration by hand; rather than replying on Log4J plugin system (so we could experiment with removing).

jodygarnett replied:
Okay if I adjust visibility I can remove the custom levels boilerplate from the xml https://osgeo-org.atlassian.net/browse/GEOT-7133

aaime commented:
Used a copy of the "release" data directory for interactive testing. Here are some findings.

- As others reported, the startup log is pretty empty, list of loaded layers/store/styles missing compared to a normal startup, despite the log stating ``Log4j 2 configuration set to DEFAULT_LOGGING.xml``.  Switching to a "GeoServer developer logging to get that information" would be overkill. Related to the "config" logging level I assume.

- Right after startup, Config UI does not show a selection of logging level, despite logging indicating the DEFAULT_LOGGING profile has been chosen (see above):

![image](https://user-images.githubusercontent.com/325433/166205963-2ab3c91e-bf34-4359-84aa-387eb321cf7b.png)

Once a logging level is chosen, the choice survives a "save" and page re-open, and the UI functions normally:

![image](https://user-images.githubusercontent.com/325433/166206090-7e3f0519-ea50-4a47-8c88-87aee8cdc605.png)

* Tested enabling/disabling log to standard output, changing log location, setting it by system variable, all worked fine.

* Wanted to see what happens on downgrading back to a previous version of GeoServer (something that users tend to do, if they find out some issue in newer versions). Seems to downgrade gracefully, starts up, notifies the properties file are not there, fills them in, and works as usual.

* Made a performance comparison between old and new (since Log4J2 is supposed to be more scalable and produce less garbage). Indeed, it is Sample request, run with 16 threads over and over:

http://localhost:8080/geoserver/tiger/wms?service=WMS&version=1.1.0&request=GetMap&layers=tiger%3Apoly_landmarks&bbox=-74.047185%2C40.679648%2C-73.90782%2C40.882078&width=528&height=768&srs=EPSG%3A4326&styles=&format=image/png

GeoServer main branch:
* default logging, 250 r/s
* production logging, 270 r/s

With this PR:
* default logging: 275 r/s
* production loggin: 280 r/s

jodygarnett commented on build/qa/pmd-ruleset.xml:
```diff
@@ -81,11 +96,11 @@ GeoTools ruleset. See https://pmd.github.io/latest/pmd_userdocs_understanding_ru
+        <!-- the method to match, so let's try to use common var names for store ...       -->
         <property name="closeTargets" value="releaseConnection,store.releaseConnection,closeQuietly,closeConnection,closeSafe,store.closeSafe,dataStore.closeSafe,getDataStore().closeSafe,close,closeResultSet,closeStmt,closeFinally,JDBCUtils.close"/>
-	<property name="allowedResourceTypes" value="java.io.ByteArrayOutputStream|java.io.ByteArrayInputStream|java.io.StringWriter|java.io.CharArrayWriter|java.util.stream.Stream|java.util.stream.IntStream|java.util.stream.LongStream|java.util.stream.DoubleStream|java.io.StringReader" />	    
+    <property name="allowedResourceTypes" value="java.io.ByteArrayOutputStream|java.io.ByteArrayInputStream|java.io.StringWriter|java.io.CharArrayWriter|java.util.stream.Stream|java.util.stream.IntStream|java.util.stream.LongStream|java.util.stream.DoubleStream|java.io.StringReader" />	    
```
fixed indent.

jodygarnett commented:
This now requires https://github.com/geotools/geotools/pull/3870 to be merged; other than that I think I have addressed all feedback.

jodygarnett commented on src/community/app-schema/webservice-test/pom.xml:
```diff
@@ -67,8 +67,12 @@
         <dependency>
-            <groupId>log4j</groupId>
-            <artifactId>log4j</artifactId>
+          <groupId>org.apache.logging.log4j</groupId>
```
was able to remove these

jodygarnett commented:
> Right after startup, Config UI does not show a selection of logging level, despite logging indicating the DEFAULT_LOGGING profile has been chosen (see above)

I was able to resolve this, on initial start value was set to "DEFAULT_LOGGING.properties" (and only "DEFAULT_LOGGING" was listed as an option in the UI)

jodygarnett commented:
> Here is a first pass of a review, based on code inspection. Will follow up with comments based on interactive testing later.

Andrea I believe I have addressed all of your feedback (and some feedback from the user list testing).

aaime commented:
Ok, I'd say proceed with the merges and then the RC then?

jodygarnett commented:
> Ok, I'd say proceed with the merges and then the RC then?

That is my plan, just waiting on upstream PR to complete its checks.
